src/serlio/materials/ArnoldMaterialNode.cpp

```diff
@@ -0,0 +1,395 @@
+/**
+ * Serlio - Esri CityEngine Plugin for Autodesk Maya
+ *
+ * See https://github.com/esri/serlio for build and usage instructions.
+ *
+ * Copyright (c) 2012-2019 Esri R&D Center Zurich
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "materials/ArnoldMaterialNode.h"
+#include "materials/MaterialInfo.h"
+#include "materials/MaterialUtils.h"
+
+#include "utils/MELScriptBuilder.h"
+#include "utils/MayaUtilities.h"
+
+#include "serlioPlugin.h"
+
+#include "maya/MFnTypedAttribute.h"
+#include "maya/adskDataStream.h"
+
+#include <mutex>
+
+namespace {
+
+const std::wstring MATERIAL_BASE_NAME = L"serlioArnoldMaterial";
+
+std::once_flag pluginDependencyCheckFlag;
+const std::vector<std::string> PLUGIN_DEPENDENCIES = {"mtoa"};
+
+void setUvTransformAttrs(MELScriptBuilder& sb, const MELVariable& node, const std::wstring& uvSet,
+                         const MaterialTrafo& trafo) {
+	sb.setAttr(node, L"uvset", L"\"" + uvSet + L"\"");
+	sb.setAttr(node, L"pivotFrame", 0.0, 0.0);
+	sb.setAttr(node, L"scaleFrame", 1.0 / trafo.su(), 1.0 / trafo.sv());
+	sb.setAttr(node, L"translateFrame", -trafo.tu() / trafo.su(), -trafo.tv() / trafo.sv());
+	if (trafo.rw() != 0.0) {
+		LOG_WRN << "rotation (material.map.rw) is not yet supported\n";
+	}
+}
+
+MELVariable createMapShader(MELScriptBuilder& sb, const std::string& mapFile, const MaterialTrafo& mapTrafo,
+                            const std::wstring& shaderName, const std::wstring& uvSet, const bool raw,
+                            const bool alpha) {
+	MELVariable mapNode(L"mapNode");
+	sb.setVar(mapNode, shaderName);
+	sb.setVar(MELVariable(L"mapFile"), prtu::toUTF16FromOSNarrow(mapFile));
+	sb.createTextureShadingNode(mapNode);
+	sb.setAttr(mapNode, L"fileTextureName", L"$mapFile");

```

Comment:
`$mapFile` is the last standing MEL variable in string representation, consider using a `MELVariable` here as well

Reply:
done